### PR TITLE
Better support for pb2 "repeated" and "message" fields.

### PR DESCRIPTION
**Pull request by lukesneeringer**
This builds on #183 by adding the ability to set within `repeated` (read: `list`) and custom (read: `dict`) fields.

Before this PR, trying to set a repeated field worked properly with Python primitives, but failed for messages.

### CODE REVIEW

File: google/gax/utils/protobuf.py [hunk 1]
```diff
@@ -124,7 +124,27 @@ def set(pb_or_dict, key, value):
     # with.
     if isinstance(pb_or_dict, collections.MutableMapping):
         pb_or_dict[key] = value
-    else:  # isinstance(pb_or_dict, Message):
+    elif isinstance(value, (collections.MutableSequence, tuple)):
+        # Clear the existing repeated protobuf message of any elements
+        # currently inside it.
+        while len(getattr(pb_or_dict, key)) > 0:
```
Comment (geigerj) [inline google/gax/utils/protobuf.py]: Can you use [`ClearField`](https://developers.google.com/protocol-buffers/docs/reference/python-generated#repeated-fields)?

File: google/gax/utils/protobuf.py [hunk 2]
```diff
@@ -124,7 +124,27 @@ def set(pb_or_dict, key, value):
...
+            getattr(pb_or_dict, key).pop()
+
+        # Write our new elements to the repeated field.
+        for v in value:
+            if isinstance(v, collections.Mapping):
+                getattr(pb_or_dict, key).add(**v)
+            else:
+                getattr(pb_or_dict, key).extend([v])
```
Comment (geigerj) [inline google/gax/utils/protobuf.py]: nit: `getattr(pb_or_dict, key).append(v)`

File: google/gax/utils/protobuf.py (hunk 1 above)
Comment (lukesneeringer) [inline google/gax/utils/protobuf.py]:
```E           AttributeError: 'google.protobuf.pyext._message.RepeatedCompositeCo' object has no attribute 'ClearField'
```

File: google/gax/utils/protobuf.py (hunk 2 above)
Comment (lukesneeringer) [inline google/gax/utils/protobuf.py]:
```
E                   AttributeError: 'google.protobuf.pyext._message.RepeatedCompositeCo' object has no attribute 'append'
```

File: google/gax/utils/protobuf.py (hunk 1 above)
Comment (geigerj) [inline google/gax/utils/protobuf.py]: I think it's [`pb_or_dict.ClearField(key)`](https://developers.google.com/protocol-buffers/docs/reference/python/google.protobuf.message.Message-class)?

**Comment by geigerj** [comment]
@xfxyjwf Could you give this a look and confirm if this looks OK from the protobuf side?

**Comment by lukesneeringer** [comment]
I will hold the PR open just in case @xfxyjwf gets back to us quickly, but if not, definitely open to amending.

**Comment by lukesneeringer** [comment]
Merging. Happy to revisit.

**Pull request merged**